### DISCUSSION TITLE
Custom instruments?

### DISCUSSION THREAD

**Discussion by agr** [open-telemetry/opentelemetry-dotnet#4461] [Q&A] [answered]
I have an Azure Function with a timer trigger that does "*something useful*". While doing that something it produces several numbers indicating the health of the process it interacts with, so I want to produce metrics out of those values. The values are useful by themselves individually. I don't need to add them, I don't want to see any histograms with it. Gauge feels like the right type of metric for those, but unfortunately there is only `ObservableGauge` which wants a callback that would return those values. I don't have a continuous process that I could sample at any moment. I have a function that works and at some point produces a few numbers. At that point (and only at that point) I want metric samples produced.

I tried to create my own `Instrument<T>` implementation, but it seems whatever samples I produce with that are just dropped. Do I have to do some additional setup besides inheriting from `Instrument<T>` and calling `Publish()` in constructor? "Standard" instruments work fine (both standard and my instruments use the same `Meter` object).

In Application Insights .net API terms, I am looking for equivalent of [`TelemetryClient.TrackMetric`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.applicationinsights.telemetryclient.trackmetric?view=azure-dotnet#microsoft-applicationinsights-telemetryclient-trackmetric(system-string-system-double-system-collections-generic-idictionary((system-string-system-string)))) function. How do I do that?

**Comment by cijothomas** [comment, marked as answer]
> I tried to create my own Instrument<T> implementation, but it seems whatever samples I produce with that are just dropped. Do I have to do some additional setup besides inheriting from Instrument<T> and calling Publish() in constructor? 

OpenTelemetry does not support any newly created instruments. It only supports the ones specified in OTel specification, and adding a new one has to go through OpenTelemetry specification process. (There were discussions around sync version of Gauge earlier, but no actual spec is made to support that yet)

**Reply by agr** [reply]
Should the `Instrument` class be marked as `internal` then?

**Reply by cijothomas** [reply]
`Instrument` is not part of this repo. Its part of DiagnosticSource package from .NET runtime, and it needs to be public to enable building OpenTelemetry instruments in this repo.
